src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ParameterBinder.cs

```diff
@@ -279,60 +224,49 @@ public class ParameterBinder
 
             var modelBindingResult = modelBindingContext.Result;
 
-            if (_validatorForBackCompatOnly != null)
-            {
-                // Since we don't have access to an IModelValidatorProvider, fall back
-                // on back-compatibility logic. In this scenario, top-level validation
-                // attributes will be ignored like they were historically.
-                if (modelBindingResult.IsModelSet)
-                {
-                    _validatorForBackCompatOnly.Validate(
-                        actionContext,
-                        modelBindingContext.ValidationState,
-                        modelBindingContext.ModelName,
-                        modelBindingResult.Model);
-                }
-            }
-            else
-            {
-                Logger.AttemptingToValidateParameterOrProperty(parameter, modelBindingContext);
+            Logger.AttemptingToValidateParameterOrProperty(parameter, modelBindingContext);
 
-                EnforceBindRequiredAndValidate(
-                    actionContext,
-                    metadata,
-                    modelBindingContext,
-                    modelBindingResult);
+            Validate(actionContext, metadata, modelBindingContext, modelBindingResult);
 
-                Logger.DoneAttemptingToValidateParameterOrProperty(parameter, modelBindingContext);
-            }
+            Logger.DoneAttemptingToValidateParameterOrProperty(parameter, modelBindingContext);
 
             return modelBindingResult;
         }
 
-        private void EnforceBindRequiredAndValidate(ActionContext actionContext, ModelMetadata metadata, ModelBindingContext modelBindingContext, ModelBindingResult modelBindingResult)
+        private void Validate(
+            ActionContext actionContext,
+            ModelMetadata metadata,
+            ModelBindingContext modelBindingContext,
+            ModelBindingResult modelBindingResult)
         {
-            if (!modelBindingResult.IsModelSet && metadata.IsBindingRequired)
+            // Validate top level properties only if users inherited from this base type,
+            // otherwise fallback to earlier behavior.
+            if (_objectModelValidator is ObjectValidatorBase validator)
```

Comment:
Likely not a big deal but I suggest storing `validator` in another field and just doing a `null`-check here.